### PR TITLE
Add annotations for classes extending `ArrayAccess`

### LINKED ISSUE #65: Improve types for `WP_REST_Request`
Right now in my project I have a bunch of REST controllers that take some request params and then do something with them. With `WP_REST_Request` implementing `ArrayAccess` and things like `$request['foo']` being `mixed`, I need to use inline `@var` tags to tell PHPStan what type I am expecting.

Example:

```php
/**
 * @param WP_REST_Request $request Full details about the request.
 * @return WP_REST_Response|WP_Error Response object on success, WP_Error object on failure.
 */
public function create_item( $request ) {
	/**
	 * Parent post.
	 *
	 * @var int $parent_post
	 */
	$parent_post = ! empty( $request['post'] ) ? $request['post'] : null;

	/**
	 * Field to order by.
	 *
	 * @var string $orderby
	 */
	$orderby = ! empty( $request['orderby'] ) ? $request['orderby'] : 'date';

	// ...
}
```

That works fine so far.

However, [with PHPStan 1.10 there is an effort underway](https://phpstan.org/blog/phpstan-1-10-comes-with-lie-detector) to get rid of these inline `@var` tags. The suggestion is to fix the wrong 3P docblocks instead of having to provide these inline tags myself.

So how can we fix the `WP_REST_Request` docblocks? I am asking this here because this can be done at the stub level and there's no need for a custom PHPStan extension.

According to [this post](https://phpstan.org/blog/generics-in-php-using-phpdocs) we should be able to add something like this to `WP_REST_Request` in the stubs:

```php
/**
 * @template T
 * @implements Collection<T>
 */
```

Note: the same goes for `WP_Theme`, which also implements `ArrayAccess`

### PR DESCRIPTION
**Pull request by swissspidy**
Covers the following classes:

* `WP_REST_Request` (exact template depends on the controller's schema)
* `WP_Theme` (the list of supported keys is hardcoded, so this could even be made more accurate)
* `WP_Block_List` (supposed to always be a list of `WP_Block` instances, for inner blocks)

Does not cover `WP_Hook`, which implements `ArrayAccess` [for legacy reasons](https://make.wordpress.org/core/2016/09/08/wp_hook-next-generation-actions-and-filters/). Figured it's not worth adding it.

Basically, this changes the stubs like so:

```php
/**
 * @phpstan-template T of array
 * @phpstan-implements ArrayAccess<key-of<T>, value-of<T>>
 */
#[\AllowDynamicProperties]
class WP_REST_Request implements \ArrayAccess
{

	// ...

	/**
	 * Retrieves a parameter from the request.
	 *
	 * @since 4.4.0
	 *
	 * @param string $offset Parameter name.
	 * @return mixed|null Value if set, null otherwise.
	 *
	 * @phpstan-param key-of<T> $offset
	 * @phpstan-return T[key-of<T>]
	 */
	#[\ReturnTypeWillChange]
	public function offsetGet($offset)
	{
	}

	// ...
}
```

And userland code can leverage it like so:

```php
/**
 * @param WP_REST_Request $request Full details about the request.
 * @return WP_REST_Response|WP_Error Response object on success, WP_Error object on failure.
 *
 * @phpstan-param WP_REST_Request<array{post?: int, orderby?: string}> $request
 */
public function create_item( $request ) {

	// This is an int.
	$parent_post = ! empty( $request['post'] ) ? $request['post'] : null;

	// This is a string.
	$orderby = ! empty( $request['orderby'] ) ? $request['orderby'] : 'date';
}
```

Fixes #65

### CODE REVIEW

**Comment by szepeviktor** [comment]
You can see the diff in CI output: https://app.travis-ci.com/github/php-stubs/wordpress-stubs/jobs/596849026

**Comment by szepeviktor** [comment]
One tiny step away from merging.

**Comment by szepeviktor** [comment]
Oh noooooooooooooooooo!
`($node instanceof Class_ && isset($node->name->name))`

**Comment by swissspidy** [comment]
> Oh noooooooooooooooooo! `($node instanceof Class_ && isset($node->name->name))`

That doesn't please PHPStan 👀

**Comment by swissspidy** [comment]
By the way, I'd appreciate feedback on the specific params for these classes.

cc @johnbillion

**Comment by szepeviktor** [comment]
Was it okay without `$node instanceof Class_`?

**Comment by swissspidy** [comment]
> Was it okay without `$node instanceof Class_`?

Yes this passes:

```php
        if (($node instanceof Function_) || ($node instanceof ClassMethod) || (isset($node->name->name))) {
            return $node->name->name;
        }
```

This does not:

```php
        if (($node instanceof Function_) || ($node instanceof ClassMethod) || ($node instanceof Class_ && isset($node->name->name))) {
            return $node->name->name;
        }
```

**Comment by szepeviktor** [comment]
How about `if (($node instanceof Function_) || ($node instanceof ClassMethod))`?

**Comment by swissspidy** [comment]
That does not match `Class_` instances though.

What about this:

```php
        if (($node instanceof Function_) || ($node instanceof ClassMethod)) {
            return $node->name->name;
        }

        if ($node instanceof Class_ && isset($node->name->name)) {
            return $node->name->name;
        }
```

That passes

File: visitor.php [hunk 1]
```diff
@@ -305,8 +306,8 @@ public function enterNode(Node $node)
 
     private static function getNodeName(Node $node): string
     {
-        if (($node instanceof Function_) || ($node instanceof ClassMethod)) {
-            return $node->name->name;
+        if (($node instanceof Function_) || ($node instanceof ClassMethod) || ($node instanceof Class_)) {
+            return $node->name->name ?? '';
```
Comment (szepeviktor) [inline visitor.php:310]: Here we go 🐎 
```suggestion
        if (($node instanceof Function_) || ($node instanceof ClassMethod) || ($node instanceof Class_ && isset($node->name))) {
            return $node->name->name;
```

Comment (swissspidy) [reply visitor.php:310]: `Cannot access property $name on PhpParser\Node\Identifier|null. `

Comment (szepeviktor) [reply visitor.php:310]: I need to understand how it works.
A PHPStan "OK" may be a bug!

Comment (szepeviktor) [reply visitor.php:310]: Trying to find the root cause with `\PHPStan\dumpType();`

Comment (szepeviktor) [reply visitor.php:310]: `&& !$node->isAnonymous()` would go for brave&potent boys but I'm looking for a solution to satisfy static analysis too.

Comment (szepeviktor) [reply visitor.php:310]: `isAnonymous` is inside phpstan.phar so cannot be analyzed. Whyyy???

Comment (herndlm) [reply visitor.php:310]: congratulations, you might have found a PHPStan bug 😅
https://phpstan.org/r/e29b76ac-5398-4505-b64a-b568c9c37c53

Comment (szepeviktor) [reply visitor.php:310]: Reported in https://github.com/phpstan/phpstan/issues/8968

**Comment by johnbillion** [comment]
@swissspidy My main question about this change is, does it work? I spent some time last year investigating how to type array access elements and didn't get anywhere. From my understanding, `@phpstan-implements ArrayAccess<key-of<T>, value-of<T>>` means that any given element can have any key from `T` and any type from `T`, without the type of the value corresponding to the key. So the type for every element is a compound of all available types.

Here's a PHPStan playground example: https://phpstan.org/r/f6e6652d-9a04-468c-aad2-2e0728112adf

**Comment by johnbillion** [comment]
Short discussion here: https://github.com/phpstan/phpstan/discussions/7062

**Comment by swissspidy** [comment]
@johnbillion I see. Haven't tested that specific scenario, will check it out tomorrow. Instead of `key-of` and `value-of`, something like this could be useful for `WP_REST_Request`: https://phpstan.org/r/4901b911-ddd4-449a-a417-2a1ec41badaa. For `WP_Theme` the list could be hardcodded. For `WP_Block_List` probably too. WDYT?

**Comment by johnbillion** [comment]
To be honest I'm not sure. This might be worth creating a discussion on the PHPStan repo.

**Comment by swissspidy** [comment]
Got something working here I think: https://phpstan.org/r/1720aa00-6cd2-48d1-b1f5-bffd59518d54

File: visitor.php (hunk 1 above)
Comment (swissspidy) [reply visitor.php]: Changing to

```php
        if ((($node instanceof Function_) || ($node instanceof ClassMethod) || ($node instanceof Class_)) && $node->name !== null) {
            return $node->name->name;
        }
```

File: visitor.php [hunk 2]
```diff
@@ -305,7 +306,7 @@ public function enterNode(Node $node)
 
     private static function getNodeName(Node $node): string
     {
-        if (($node instanceof Function_) || ($node instanceof ClassMethod)) {
+        if ((($node instanceof Function_) || ($node instanceof ClassMethod) || ($node instanceof Class_)) && $node->name !== null) {
```
Comment (szepeviktor) [inline visitor.php:309]: Please write [positive conditions](https://github.com/szepeviktor/very-clean-code/blob/master/README.md#-write-positive-conditions).

**Comment by szepeviktor** [comment]
> Stay positive

👍🏻

It is still a Draft?

**Comment by swissspidy** [comment]
Yes. Doing some more testing tomorrow before marking as ready for review. Not 100% happy with the `WP_Theme` annotations yet. Maybe I'll remove them and just keep `WP_REST_Request` for now, because that is the most useful one.

**Comment by swissspidy** [comment]
I'd say the current definitions are good enough and they can be improved later if needed

**Comment by szepeviktor** [comment]
Okay. Jumping into an empty swimming pool.

**Pull request merged**